# `pandas`

This workshop's goal&mdash;which is facilitated by this Jupyter notebook&mdash;is to give attendees the confidence to use `pandas` in their research projects. Basic familiarity with Python and with Jupyter Notebooks *is* assumed.

`pandas` is designed to make it easier to work with structured data. Most of the analyses you might perform will likely involve using tabular data, e.g., from .csv files or relational databases (e.g., SQL). The `DataFrame` object in `pandas` is "a two-dimensional tabular, column-oriented data structure with both row and column labels."

If you're curious:

>The `pandas` name itself is derived from *panel data*, an econometrics term for multidimensional structured data sets, and *Python data analysis* itself. You can consult the full `pandas` documentation [here](http://pandas.pydata.org/pandas-docs/stable/).

To motivate this workshop, we'll work with example data and go through the various steps you might need to prepare data for analysis. You'll (hopefully) realize that doing this type of work is much more difficult using Python's built-in data structures.

### Table of Contents

1 - [The DataFrame](#section1)<br>

2 - [Rename, Index, and Slice](#section2)<br>

3 - [Manipulating Columns](#section3)<br>

4 - [Merging](#section4)<br>

5 - [Calculating Unique And Missing Values](#section5)<br>

6 - [Groupby](#section6)<br>

7- [Exporting To CSV](#section7)<br>

8 - [Handling Missing Values (Boolean Indexing)](#section8)<br>

9 - [Sorting Values](#section9)<br>

10 - [Plotting In Pandas](#section10)<br>

## 1. The DataFrame: Importing Data and Summary Statistics <a id="section1"/>
The data used in these examples is available in the following [GitHub repository](https://github.com/dlab-berkeley/introduction-to-pandas). If you've [cloned that repo](https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-clone), which is the recommended approach, you'll have everything you need to run this notebook. Otherwise, you can download the data file(s) from the above link. (Note: this notebook assumes that the data files are in a directory named `data/` found within your current working directory.)

For this example, we're working with European unemployment data from Eurostat, which is hosted by [Google](https://code.google.com/p/dspl/downloads/list). There are several .csv files that we'll work with in this workshop.

Let's begin by importing `pandas` using the conventional abbreviation `pd`.

In [1]:
%matplotlib inline

import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt

mpl.rc('savefig', dpi=200)
plt.style.use('ggplot')
plt.rcParams['xtick.minor.size'] = 0
plt.rcParams['ytick.minor.size'] = 0

The `read_csv()` function in `pandas` allows us to easily import our data. By default, it assumes the data is comma-delimited. However, you can specify the delimiter used in your data (e.g., tab, semicolon, pipe, etc.). There are several other parameters that you can specify, if needed. See the documentation [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). `read_csv()` returns a `DataFrame`.

Notice that we call `read_csv()` using the `pd` abbreviation from the import statement above.

In [2]:
unemployment = pd.read_csv('data/country_total.csv')

Great! You've created a `pandas` `DataFrame`, which is now stored in the variable `unemployement`. We can look at our data by using the `.head()` method. By default, this shows the header (column names) and the first five rows. Passing an integer, $n$, to `.head()` returns that number of rows.

In [3]:
unemployment.head()

country seasonality    month  unemployment  unemployment_rate
0      at         nsa  1993.01        171000                4.5
1      at         nsa  1993.02        175000                4.6
2      at         nsa  1993.03        166000                4.4
3      at         nsa  1993.04        157000                4.1
4      at         nsa  1993.05        147000                3.9

To find the number of rows, you can use the `len()` function. Alternatively, you can use the `shape` attribute, which returns a tuple with the number of rows and columns.

In [4]:
len(unemployment)

20796

In [5]:
unemployment.shape

(20796, 5)

There are 20,796 rows and 5 columns in your `unemployment` DataFrame.

A useful method that generates various summary statistics is `.describe()`.

In [6]:
unemployment.describe()

month  unemployment  unemployment_rate
count  20796.000000  2.079600e+04       19851.000000
mean    1999.401290  7.900818e+05           8.179764
std        7.483751  1.015280e+06           3.922533
min     1983.010000  2.000000e+03           1.100000
25%     1994.090000  1.400000e+05           5.200000
50%     2001.010000  3.100000e+05           7.600000
75%     2006.010000  1.262250e+06          10.000000
max     2010.120000  4.773000e+06          20.900000

You may have noticed that the "count" is lower for the unemployment rate. This is because the summary statistics are based on *non-missing* values and count reflects that.

The values depend on what it's called on. If the `DataFrame` includes both numeric and object (e.g., strings) `dtype`s like in our case, it will default to summarizing the numeric data. If `.describe()` is called on strings, for example, it will return the count, number of unique values, and the most frequent value along with its count.

You can use the `.info()` method to have a look at the `dtype` of each column and at how many *non-missing* values they contain.

In [7]:
unemployment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20796 entries, 0 to 20795
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   country            20796 non-null  object 
 1   seasonality        20796 non-null  object 
 2   month              20796 non-null  float64
 3   unemployment       20796 non-null  int64  
 4   unemployment_rate  19851 non-null  float64
dtypes: float64(2), int64(1), object(2)
memory usage: 812.5+ KB


### Challenge 1: Import Data From A URL

Above, we imported the unemployment data using the `read_csv` function and a relative file path. `read_csv` is [a very flexible method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.read_csv.html); it also allows us to import data using a URL as the file path. 

A csv file with data on world countries and their abbreviations is located at [https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv](https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv) (saved as a string variable `countries_url` below).

Using `read_csv`, import the country data and save it to the variable `countries`.

In [8]:
countries_url = 'https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv'
countries = ...

### Challenge 2: The `tail` method

DataFrames all have a method called `tail` that takes an integer as an argument and returns a new DataFrame. Before using `tail`, can you guess at what it does? Try using `tail`; was your guess correct?

In [9]:
# use the tail method on either the unemployment or countries dataframe


### Challenge 3: Describe `countries`

It's important to understand a few fundamentals about your data before you start work with it, including what information it contains, how large it is, and how the values are generally distributed.

Using the methods and attributes above, answer the following questions about `countries`:
* what columns does it contain?
* what does each row stand for?
* how many rows and columns does it contain?
* are there any missing values in the latitude or longitude columns? 

Hint: the `head` and `describe` functions, as well as the `shape` attribute, will be helpful here.

In [10]:
# explore the countries data


**YOUR ANSWER HERE**

## 2. Rename, Index, and Slice <a id="section2"/>
Back to the entire unemployment data set. You may have noticed that the `month` column also includes the year. Let's go ahead and rename it.

In [11]:
unemployment.rename(columns={'month' : 'year_month'}, inplace=True)

The `.rename()` method allows you to modify column names and/or column names index labels. As you can see, we passed a `dict` to the `columns` parameter, with the original name as the key and the new name as the value. Importantly, we also set the `inplace` parameter to `True`, which modifies the *actual* `DataFrame`, not a copy of it.

It might also make sense to separate the data in `year_month` into two separate columns. To do this, you'll need to know how to select a single column first. We can either use bracket (`[]`) or dot notation (referred to as *attribute access*).

In [12]:
unemployment['year_month'].head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

In [13]:
unemployment.year_month.head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

It is preferrable to use the bracket notation as a column name might inadvertently have the same name as a `DataFrame` (or `Series`) method. In addition, only bracket notation can be used to create a new column. If you try and use attribute access to create a new column, you'll create a new attribute, *not* a new column. We will use the bracket notation throughout this Notebook.

When selecting a single column, we have a `pandas` `Series` object, which is a single vector of data (e.g., a NumPy array) with "an associated array of data labels, called its *index*. A `DataFrame` also has an index. In our example, the indices are an array of sequential integers, which is the default. You can find them in the left-most position, in bold and without a column label.

Indices need not be a sequence of integers. They can, for example, be dates or strings. Note that indices do *not* need to be unique.

Indices, like column names, can be used to select data. Indices can be used to select particular rows. In fact, you can do something like `.head()` with slicing using the `[]` operator.

In [14]:
unemployment[:5]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
1      at         nsa     1993.02        175000                4.6
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1
4      at         nsa     1993.05        147000                3.9

Before we continue, let's look at a few useful ways to index data&mdash;that is, select rows.

`.loc` primarily works with string labels. It accepts a single label, a list (or array) of labels, or a slice of labels (e.g., `'a' : 'f'`) as argument.

Let's create a `DataFrame` to see how this works. (This is based on an [example](https://github.com/fonnesbeck/scipy2015_tutorial/blob/master/notebooks/1.%20Data%20Preparation.ipynb) from Chris Fonnesbeck's [Computational Statistics II Tutorial](https://github.com/fonnesbeck/scipy2015_tutorial).)

In [15]:
bacteria = pd.DataFrame({'bacteria_counts' : [632, 1638, 569, 115],
                         'other_feature' : [438, 833, 234, 298]},
                         index=['Firmicutes', 'Proteobacteria', 'Actinobacteria', 'Bacteroidetes'])

Notice that we pass in a `dict`, where the keys correspond to column names and the values to the data. In this example, we've also set the indices&mdash;strings in this case&mdash;to be the taxon of each bacterium.

In [16]:
bacteria

bacteria_counts  other_feature
Firmicutes                  632            438
Proteobacteria             1638            833
Actinobacteria              569            234
Bacteroidetes               115            298

Now, if we're interested in the values (row) associated with "Actinobacteria," we can use `.loc` and the index name.

In [17]:
bacteria.loc['Actinobacteria']

bacteria_counts    569
other_feature      234
Name: Actinobacteria, dtype: int64

This returns the column values for the specified row. Interestingly, we could have also used "positional indexing," even though the indices are strings.

In [18]:
bacteria[2:3]

bacteria_counts  other_feature
Actinobacteria              569            234

The difference is that using `.loc` returns a `Series` because we selected a single label, while using the positional indexing returns a `DataFrame` because we selected a range of positions.

Let's return to our unemployment data. Another indexing option, `.iloc`, primarily works with integer positions. To select specific rows, we can pass it a list of integers.

In [19]:
unemployment.iloc[[1, 5, 6, 9]]

country seasonality  year_month  unemployment  unemployment_rate
1      at         nsa     1993.02        175000                4.6
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
9      at         nsa     1993.10        141000                3.7

We can also select a range of rows and specify the step value.

In [20]:
unemployment.iloc[25:50:5]

country seasonality  year_month  unemployment  unemployment_rate
25      at         nsa     1995.02        174000                4.5
30      at         nsa     1995.07        123000                3.3
35      at         nsa     1995.12        175000                4.7
40      at         nsa     1996.05        159000                4.3
45      at         nsa     1996.10        146000                3.9

(Note: As is typical in Python, the end position is not included. Therefore, we don't see the row associated with the index 50.)

Indexing is important, you'll use it a lot. Below, we'll show how to index based on data values.



### Challenge 4: Renaming `bacteria`

The "other_feature" column in our `bacteria` table isn't very descriptive. Suppose we know that "other_feature" refers to a second set of bacteria count observations. Use the `.rename` method to give "other_feature" a more descriptive name.

In [21]:
# rename "other_feature" in bacteria


### Challenge 5: Indexing to get a specific value

Both `.loc` and `.iloc` can be used to select a particular value if they are given two arguments. The first argument is the name (when using `.loc`) or index number (when using `.iloc`) of the *row* you want, while the second argument is the name or index number of the *column* you want.

Using `.loc`, select "Bacteroidetes" and "bacteria_counts" to get the count of Bacteroidetes.

BONUS: how could you do the same task using `.iloc`?

### Challenge 6: Indexing multiple rows and columns

Both `.loc` and `.iloc` can be used to select subsets of columns *and* rows at the same time if they are given lists (and/or slices, for `.iloc`) as their two arguments. 

Using `.iloc` on the `unemployment` DataFrame, get:
* every row starting at row 4 and ending at row 7
* the 0th, 2nd, and 3rd columns

BONUS: how could you do the same task using `.loc`?

## 3. Manipulating Columns: Renaming, Creating, Reordering

So, we still want to **split `year_month` into two separate columns.** Above, we saw that this column is type (technically, `dtype`) `float64`. We'll first extract the year using the `.astype()` method. This allows for type casting&mdash;basically converting from one type to another. We'll then subtract this value from `year_month`&mdash;to get the decimal portion of the value&mdash;and multiply the result by 100 and convert to `int`.

For more information on `pandas` `dtype`s, check the documentation [here](http://pandas.pydata.org/pandas-docs/stable/basics.html#dtypes).

In [22]:
unemployment['year'] = unemployment['year_month'].astype(int)

In this case, we're casting the floating point values to integers. In Python, this [truncates the decimals](https://docs.python.org/2/library/stdtypes.html#numeric-types-int-float-long-complex).

Finally, let's create our **month** variable as described above. (Because of the truncating that occurs when casting to `int`, we first round the values to the nearest whole number.)

In [23]:
unemployment['month'] = ((unemployment['year_month'] - unemployment['year']) * 100).round(0).astype(int)

In [24]:
unemployment.head(12)

country seasonality  year_month  unemployment  unemployment_rate  year  \
0       at         nsa     1993.01        171000                4.5  1993   
1       at         nsa     1993.02        175000                4.6  1993   
2       at         nsa     1993.03        166000                4.4  1993   
3       at         nsa     1993.04        157000                4.1  1993   
4       at         nsa     1993.05        147000                3.9  1993   
5       at         nsa     1993.06        134000                3.5  1993   
6       at         nsa     1993.07        128000                3.4  1993   
7       at         nsa     1993.08        130000                3.4  1993   
8       at         nsa     1993.09        132000                3.5  1993   
9       at         nsa     1993.10        141000                3.7  1993   
10      at         nsa     1993.11        156000                4.1  1993   
11      at         nsa     1993.12        169000                4.4  1993   

    month  
0       1  
1       2  
2       3  
3       4  
4       5  
5       6  
6       7  
7       8  
8       9  
9      10  
10     11  
11     12

To create the `month` column, we subtracted two vectors. This resulted in the decimal value in `year_month`. To transform the values to integers, we multiplied by 100.

Now, let's say we wanted to **reorder the columns** in the `DataFrame`. For this, we use bracket notation again, passing in a list of column names in the order we'd like to see them.

In [25]:
unemployment = unemployment[['country', 'seasonality',
                             'year_month', 'year', 'month',
                             'unemployment', 'unemployment_rate']]

In [26]:
unemployment.head(10)

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   
5      at         nsa     1993.06  1993      6        134000   
6      at         nsa     1993.07  1993      7        128000   
7      at         nsa     1993.08  1993      8        130000   
8      at         nsa     1993.09  1993      9        132000   
9      at         nsa     1993.10  1993     10        141000   

   unemployment_rate  
0                4.5  
1                4.6  
2                4.4  
3                4.1  
4                3.9  
5                3.5  
6                3.4  
7                3.4  
8                3.5  
9                3.7

### Challenge 7: Another way to get the month

If you didn't know that casting floats to ints truncates the decimals in Python, you could have used NumPy's `floor()` function. `np.floor` takes an array or Pandas Series of floats as its argument, and returns an array or Series where every float has been rounded down to the nearest whole number. 


Use `np.floor` to round the values in the "year_month" column down so we can cast them as integer years. Note that the types are still floats, so we'll still need to use `astype` to typecast.

In [27]:
import numpy as np

# select the "year_month" column
year_month = ...

# use np.floor on year_month to get the years as floats
years_by_floor = ...

# cast years_by_floor to integers using astype(int)
int_years = ...

# check that this gets the same answers as our first approach
# this should return True
(unemployment['year_month'].astype(int) == int_years).all()

ValueError: unknown type object

The last line of code in the previous cell does an element-wise comparison of the values in the corresponding arrays. The `.all()` method checks whether *all* elements are `True`.

## 4. Merging

So far, our `DataFrame` is organized in a reasonable way. But, we know we can do better. We're eventually going to be interested in the unemployment rate for each country. The trouble is, we don't exactly know what the values in `country` refer to. We can fix that by getting country names. These can be found in `countries.csv`. Run the following cell to import the countries data using a URL as the file path.

In [28]:
countries_url = 'https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv'
countries = pd.read_csv(countries_url)

In [29]:
countries.tail(3)

country google_country_code country_group         name_en      name_fr  \
27      se                  SE            eu          Sweden        Suède   
28      tr                  TR        non-eu          Turkey      Turquie   
29      uk                  GB            eu  United Kingdom  Royaume-Uni   

                   name_de   latitude  longitude  
27                Schweden  62.198468  14.896307  
28                  Türkei  38.952942  35.439795  
29  Vereinigtes Königreich  54.315447  -2.232612

This file has lots of useful information. It even has the country names is three different languages.

Because the data we need is stored in two separate files, we'll want to **merge** the data somehow. Let's determine which column we can use to join this data. `country` looks like a good option. However, we don't need all of the columns in the `countries` `DataFrame`. To select certain columns, we use the name bracket notation we used to reorder the columns.

In [ ]:
country_names = countries[['country', 'country_group', 'name_en']]

In [ ]:
country_names.head(2)

`pandas` includes an easy-to-use merge function. Let's use it to **merge the two `DataFrame`s on country code.**

In [ ]:
unemployment = pd.merge(unemployment, country_names, on='country')

Merging is often more complex than this example. If you want to merge on multiple columns, you can pass a list of column names to the `on` parameter.

```
pd.merge(first, second, on=['name', 'id'])
```

For more information on merging, check the [documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging).

`pandas` also provides a `.merge()` method that can act on a `DataFrame`. You can read more about that [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html).

### Challenge 8: 

You may sometimes need to merge on columns with different names. To do so, use the `left_on` and `right_on` parameters, where the first listed `DataFrame` is the "left" one and the second is the "right." It might look something this.

```
pd.merge(one, two, left_on='city', right_on='city_name')
```

Suppose wanted to merge `unemployment` with a new DataFrame called `country_codes`, where the abbreviation for each country is in the column "c_code":

In [ ]:
country_codes = country_names.rename({"country":"c_code"}, axis=1).drop("country_group", axis=1)
country_codes.head()

Use `merge` to merge `unemployment` and `country_codes` on their country codes. Make sure to specify `left_on=` and `right_on=` in the call to `merge`!

## 5. Calculating Unique and Missing Values

In [ ]:
unemployment.head()

That's better. We now know that the abbreviation "at" corresponds to Austria. We might be curious to check what countries we have data for. The `Series` object includes a `.unique()` method. We'll use this to check the countries. We can select the name either using bracket or dot notation. (While we suggested using brackets above, it *is* sometimes easier to use dot notation. Just be careful.)

In [ ]:
unemployment.name_en.unique()

To get a count of the **number of unique countries,** we can either wrap the above code with `len()` to get the number of items in the array, or we can use the  `Series.nunique()` method.

In [ ]:
unemployment.name_en.nunique()

It might be more interesting to know **how many observations** we actually have. `pandas` has a `Series` method called `.value_counts()` that returns the counts for the unique values in the `Series`.

In [ ]:
unemployment['name_en'].value_counts()

By default, the `Series` is sorted by values. If you'd like it sorted by index&mdash;country name in this case&mdash;append the `.sort_index()` method.

In [ ]:
unemployment['name_en'].value_counts().sort_index()

This will be useful for our analysis. The maximum number of observations for a given country for this time period is 1,008 observations. We'll note that certain countries, such as Turkey, have far less data.

How about finding the **date range** for this data set? Let's look at the minimum and maximum years.

In [ ]:
unemployment['year'].min(), unemployment['year'].max()

Next, we should pause for a moment and think about what data we really care about. For our purposes, the variable of interest is `unemployment_rate`. The number of observations by country only reflect the number of instances of each country name in the dataset. It is possible, maybe even expected, to have some missing data within those instances. Let's find out **how many unemployment rate values are missing.**

In [ ]:
unemployment['unemployment_rate'].isnull().sum()

The `.isnull()` method returns a corresponding boolean value for each entry in the unemployment rate `Series`. In Python `True` is equivalent to 1 and `False` is equivalent to 0. Thus, when we add the result (with `.sum()`), we get a count for the *total* number of missing values.

### Challenge 9: Exploring unemployment rates

What are the minimum and maximum unemployment rates in our data set? Which unemployment rates are most and least common?

Hint: look at where we found the minimum and maximum years for a hint to the first question, and use `value_counts` for the second.

## 6. GroupBy


What if we'd like to know how many missing values exist at the *country* level? We can take the main part of what we had above and create a new column in the `DataFrame`.

In [ ]:
unemployment['unemployment_rate_null'] = unemployment['unemployment_rate'].isnull()

To count the **number of missing values for each country,** we introduce the `.groupby()` method.

In [ ]:
unemployment.groupby('name_en')['unemployment_rate_null'].sum()

Let's explain what just happened. We start with our `DataFrame`. We tell `pandas` that we want to group the data by country name&mdash;that's what goes in the parentheses. Next, we need to tell it what column we'd like to perform the `.sum()` operation on. In this case, it's the indicator for whether or not the unemployment rate was missing.

As we saw above, the number of records for each country differs. We might, then, want to have the **missing values by country shown as percentages.** Let's create a new `DataFrame` for this.

We'll take the code from above and set the `as_index` parameter to `False`.

In [ ]:
unemployment_rate = unemployment.groupby('name_en', as_index=False)['unemployment_rate_null'].sum()

In [ ]:
unemployment_rate.head(3)

`unemployment_rate` is a `DataFrame` with the information from cell 34, above. It's important to note that using `as_index=False` in `.groupby()` only works if the grouping column(s) are not the same as the columns on which we're performing the operation.

Also, to group by several columns, simply pass in a list of column names to `.groupby()`.

```
unemployment.groupby(['name_en', 'seasonality'])['unemployment_rate'].mean()
```

Now, let's add the number of observations by country to the `DataFrame`.

In [ ]:
unemployment_rate['n_obs'] = unemployment.groupby('name_en')['name_en'].count().values

Here, we need to use the `values` attribute to get an array of the counts. Excluding `values` will result in a column full of `NaN`s. This is because the index in `unemployment.groupby('name_en')['name_en'].count()` is a list of the country names. When creating a new column, `pandas` tries to match on index. Recall that the default index values for a `DataFrame` is a sequence of integers.

Because we know (or have noticed) that the `.groupby()` function returns the values in alphabetical order, we can simply set the new column to the list of values, as we have done. You can, however, be more explicit and create another `DataFrame` and merge on country name.

Finally, let's create the column for the percentage of missing values.

In [ ]:
unemployment_rate['null_percentage'] = unemployment_rate['unemployment_rate_null'] / unemployment_rate['n_obs']

In [ ]:
unemployment_rate

This is the second time we've called a `DataFrame` without something like `.head()`. Let's describe what it does. By default, `pandas` prints 60 rows. In this case, because there are only 30 countries, we see the entire `DataFrame`.

### Challenge 10: GroupBy 

Find the average unemployment rate for European Union vs. non-European Union countries. 

1. use `groupby` to group on "country_group"
2. select the "unemployment_rate" column
3. use `.mean()` to get the average

## 7. Exporting A DataFrame to csv <a id="section7"/>

As we can see, Croatia has lots of missing data. This `DataFrame` contains useful information&mdash;things to consider&mdash;when analyzing the data.

Suppose we wanted to save this as a .csv file. For this, we'd use the `.to_csv()` method.

In [ ]:
unemployment_rate.to_csv('data/unemployment_missing.csv')

Let's look at our file.

In [ ]:
!head -5 data/unemployment_missing.csv

By default, this method writes the indices. We probably don't want that. Let's edit the code. Let's also be explicit about the type of delimiter we're interested in. (Values can be separated by pipes (`|`), semicolons (`;`), tabs (`\t`), etc.)

In [ ]:
unemployment_rate.to_csv('data/unemployment_missing.csv', index=False, sep=',')

In [ ]:
!head -5 data/unemployment_missing.csv

Much better!

Let's return to our main `DataFrame`. Now that we have the missing values information in `unemployment_rate`, we can **drop the last column** we added to `unemployment`.

In [ ]:
unemployment.drop('unemployment_rate_null', axis=1, inplace=True)

It's important to specify the `axis` parameter. `axis=1` refers to columns (`axis=0` refers to rows.) The parameter `inplace=True` simply modifies the actual `DataFrame` rather than returning a new `DataFrame`.

In [ ]:
unemployment.head()

## 8. Dealing With Missing Values: Boolean Indexing <a id="section8"/>

Now that we know about the missing values, we have to deal with them. There are two main options:

* Fill the missing values with some other values.
* Do not use observations with missing values.
    * Depending on the analysis, we may want to exclude entire countries.
    
Because countries with missing unemployment rate data have at least 36 missing values, which is too many to fill, we'll take the second approach and **exclude missing values** from our primary analyses.

Instead of just getting rid of that data, it might make sense to store it in a separate `DataFrame`. This way, we could answer questions such as, "do missing values occur during certain months (or years) more frequently?" With this, we will introduce the concept of *boolean indexing* for filtering data.

In [ ]:
unemployment_rate_missing = unemployment[unemployment['unemployment_rate'].isnull()]

Recall that `unemployment['unemployment_rate'].isnull()` produces an array of boolean values. We used this previously when counting the number of missing values, though we did not see its output. Let's see some of that now.

In [ ]:
unemployment['unemployment_rate'].isnull()[:10]

To create `unemployment_rate_missing`, we're indexing `unemployment` with the array above. This returns only the rows where the value in the array is `True`. Let's see if it worked.

In [ ]:
unemployment_rate_missing.head()

It is also possible to specify multiple conditions using the `&` operator, but each condition needs to be inside of parentheses. The `.isin()` method, which takes a `list` of values, is useful when you're interested in conditioning on multiple values on a given column. For example, if you want to select multiple countries.

Now, we're ready to remove the missing data in `unemployment`. To do this, we can use the `.dropna()` method.

In [ ]:
unemployment.dropna(subset=['unemployment_rate'], inplace=True)

(Note that while we have dropped all observations for which `unemployment_rate == NaN`, this doesn't mean that all of our observations overlap exactly in time. We may find that there are dates where we have data for one country and no data for others.)

### Challenge 11: Boolean Indexing
Suppose we only want to look at unemployment data from the year 2000 or later. Use Boolean indexing to create a DataFrame with only these years.

1. Select the "year" column from `unemployment`
2. Using the year data, create a **mask**: an array of Booleans where each value is True if and only if the year is 2000 or later. Remember, you can use Boolean operators like `>`, `<`, and `==` on a column
3. Use the mask from step 2 to index `unemployment`

In [ ]:
# select the year column from unemployment
year = ...

# create a mask
later_or_equal_2000 = ...

# Boolean index unemployment
...

## 9. Sorting Values

At this point, you might be curious to know what the highest unemployment rates were. For this, we'll use the `DataFrame.sort_values()` method to **sort the data.**

In [ ]:
unemployment.sort_values('unemployment_rate', ascending=False)[:5]

The above code creates a copy of the `DataFrame`, sorted in *descending* order, and prints the first five rows.

You may have noticed that the data set includes a `seasonality` column, which we haven't yet discussed. The unemployment rate in this data is actually calculated in three separate ways. Let's look at the values.

In [ ]:
unemployment['seasonality'].unique()

The three options above correspond to:

* not seasonally adjusted
* seasonally adjusted
* trend cycle

We'll stick with seasonally adjusted data so that the values are more comparable. Let's look at the highest unemployment rates in this context.

In [ ]:
unemployment[unemployment['seasonality'] == 'sa'].sort_values('unemployment_rate', ascending=False)[:5]

Spain has the highest seasonally adjusted unemployment rate.



## 10. Plotting With Pandas

The best way to get a sense of this data is to **plot it.** Next, we'll start to look at some basic plotting with `pandas`. Before we begin, let's sort the data by country and date. This is good practice and is especially important when using `pandas`'s `.plot()` method because the x-axis values are based on the indices. When we sort, the index values remain unchanged. Thus, we need to reset them. The `drop` parameter tells `pandas` to construct a `DataFrame` *without* adding a column.

In [ ]:
unemployment.sort_values(['name_en', 'year_month'], inplace=True)
unemployment.reset_index(drop=True, inplace=True)

Let's take a look at Spain's unemployment rate (only because it was the highest) across time.

In [ ]:
spain = unemployment[(unemployment['name_en'] == 'Spain') &
                     (unemployment['seasonality'] == 'sa')]

In [ ]:
spain['unemployment_rate'].plot(figsize=(10, 8), color='#348ABD')

Note that the values along the x-axis represent the indices associated with Spain in the sorted `unemployment` `DataFrame`. Wouldn't it be nice if, instead, we could **show the time period** associated with the various unemployment rates for Spain? It might also be interesting to **compare** Spain's unemployment rate with its neighbor to the west, Portugal.

Let's first create a `DataFrame` that contains the unemployment data for both countries.

In [ ]:
ps = unemployment[(unemployment['name_en'].isin(['Portugal', 'Spain'])) &
                  (unemployment['seasonality'] == 'sa')]

Next, we'll **generate time series data** by converting our years and months into `datetime` objects. `pandas` provides a `to_datetime()` function that makes this relatively simple. It converts an argument&mdash;a single value or an array of values&mdash;to `datetime`. (Note that the return value [depends on the input](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html).) If we were interested in March 23, 1868, for example, we could do the following.

In [ ]:
pd.to_datetime('1868/3/23')

The argument doesn't necessarily have to be specified in the `yyyy/mm/dd` format. You could list it as `mm/dd/yyyy`, but it's a good idea to be explicit. As a result, we pass in a valid string format.

In [ ]:
pd.to_datetime('3/23/1868', format='%m/%d/%Y')

Let's create the `datetime` object and add it to the `DataFrame` as a column named `date`. For this, we'll use the `DataFrame.insert()` method.

In [ ]:
ps.insert(loc=0, column='date',
          value=pd.to_datetime(ps['year'].astype(str) + '/' + ps['month'].astype(str) + '/1'))

Finally, let's only keep certain columns, rename them, and reshape the `DataFrame`.

In [ ]:
ps = ps[['date', 'name_en', 'unemployment_rate']]
ps.columns = ['Time Period', 'Country', 'Unemployment Rate']
ps = ps.pivot(index='Time Period', columns='Country', values='Unemployment Rate')
ps.tail()

In [ ]:
ps.head()

Notice the indices.

In [ ]:
ps.plot(figsize=(10, 8), title='Unemployment Rate\n')

### Challenge 12: Plot without missing values
Note that there are some dates for which we lack data on Spain's unemployment rate. What could you do if you wanted your plot to show only dates where both Spain and Portugal have a defined unemployment rate?